File: backend/app/routers/trial.py
"""Xin và xem hạn ngạch dùng thử ẩn danh.

Hai endpoint, và cả hai đều công khai — phải thế, vì đây là cách một người chưa
có gì trong tay bắt đầu.

Vì sao `start` là POST chứ không phải GET
------------------------------------------
Nó ĐẶT một cookie, tức là thay đổi trạng thái. Một GET đặt cookie sẽ bị trình
duyệt nạp trước (prefetch) và bởi mọi bộ quét link, nên hạn ngạch bắt đầu tiêu
trước khi người dùng bấm gì.
"""

from __future__ import annotations

from typing import Any, Dict

from fastapi import APIRouter, Request, Response

from app import trial
from app.rate_limit import enforce_ip_limit

router = APIRouter(prefix="/trial", tags=["trial"])


def _payload(state: trial.TrialState, has_grant: bool) -> Dict[str, Any]:
    return {
        "has_grant": has_grant,
        "minutes_limit": state.minutes_limit,
        "minutes_used": state.minutes_used,
        "minutes_remaining": state.minutes_remaining,
        "resets_at": state.resets_at,
        "exhausted": not state.allowed and has_grant,
    }


@router.post("/start")
def start_trial(request: Request, response: Response) -> Dict[str, Any]:
    """Cấp phiếu dùng thử, hoặc trả lại tình trạng của phiếu đang có.

    Idempotent theo cookie: bấm hai lần không cấp phiếu thứ hai và không làm mới
    hạn ngạch. Nếu cấp mới mỗi lần gọi thì hạn ngạch hằng ngày trở thành vô hạn —
    chỉ cần gọi lại endpoint này.
    """
    # Giới hạn theo IP ở ĐÂY thì hợp lý, khác với việc dùng IP làm danh tính:
    # cái này chỉ chặn một máy đúc hàng loạt phiếu, không quyết định ai là ai.
    enforce_ip_limit(request, bucket="trial_start", max_calls=20, window=3600)

    existing = trial.peek(request)
    if existing.grant_id is not None:
        return _payload(existing, has_grant=True)

    trial.issue(response)
    # Phiếu vừa cấp: chưa tiêu phút nào, và đồng hồ chưa chạy — nó chỉ chạy ở
    # lượt suy luận đầu tiên.
    fresh = trial.TrialState(
        allowed=True, minutes_used=0,
        minutes_limit=existing.minutes_limit,
        resets_at=existing.resets_at, grant_id="new",
    )
    return _payload(fresh, has_grant=True)


@router.get("/status")
def trial_status(request: Request) -> Dict[str, Any]:
    """Số phút còn lại. Không tiêu tốn gì — giao diện gọi cái này để vẽ đồng hồ."""
    state = trial.peek(request)
    return _payload(state, has_grant=state.grant_id is not None)

File: frontend/src/api/trial.ts
import axiosClient from "./axiosClient";

/**
 * Phiếu dùng thử ẩn danh.
 *
 * Vì sao phần này phải có mặt giao diện
 * --------------------------------------
 * Cổng gác (`app/access_gate.py`) cho phép `/realtime/*` chạy với **một trong
 * hai**: phiên đăng nhập, HOẶC một phiếu dùng thử. Nhưng phiếu chỉ lấy được
 * bằng `POST /trial/start`, và trước tệp này **không có chỗ nào trong giao diện
 * gọi nó**. Hệ quả: tuyến `/realtime` mở cho khách vãng lai, khách bấm vào, và
 * mọi lời gọi API trả 401 — tính năng dùng thử tồn tại đầy đủ ở máy chủ mà
 * không ai chạm tới được.
 *
 * `start` là POST chứ không phải GET vì nó ĐẶT cookie. Một GET đặt cookie sẽ bị
 * trình duyệt nạp trước và bởi mọi bộ quét liên kết, nên hạn ngạch bắt đầu tiêu
 * trước khi người dùng bấm gì.
 */

const API_PREFIX = "/api/v1/trial";

/** Sự kiện phát ra mỗi khi máy chủ báo lại số phút còn lại. Xem `axiosClient`. */
export const TRIAL_EVENT = "voya:trial-minutes";

export interface TrialState {
  /** Đã có phiếu chưa. `false` nghĩa là chưa bấm "dùng thử" lần nào. */
  has_grant: boolean;
  minutes_limit: number;
  minutes_used: number;
  minutes_remaining: number;
  /** Mốc hạn ngạch làm mới, dạng ISO. */
  resets_at: string | null;
  /** Có phiếu nhưng đã tiêu hết phút hôm nay. */
  exhausted: boolean;
}

/**
 * Xin phiếu, hoặc lấy lại tình trạng phiếu đang có.
 *
 * Idempotent theo cookie: bấm hai lần không cấp phiếu thứ hai và **không làm
 * mới hạn ngạch**. Nếu nó cấp mới mỗi lượt gọi thì hạn ngạch hằng ngày thành vô
 * hạn — chỉ cần gọi lại đúng endpoint này.
 */
export async function startTrial(): Promise<TrialState> {
  const res = await axiosClient.post<TrialState>(`${API_PREFIX}/start`);
  return res.data;
}

/** Số phút còn lại. Không tiêu tốn gì — dùng để vẽ đồng hồ. */
export async function fetchTrialStatus(): Promise<TrialState> {
  const res = await axiosClient.get<TrialState>(`${API_PREFIX}/status`);
  return res.data;
}

/**
 * Đọc số phút còn lại từ header của MỘT phản hồi bất kỳ.
 *
 * Cổng gác gắn `X-Trial-Minutes-Remaining` vào **mọi** phản hồi đi qua phiếu,
 * không chỉ lúc hết. Nhờ vậy đồng hồ trên màn hình cập nhật theo từng lượt
 * nhận dạng mà không phải gọi thêm một vòng `/trial/status` cho mỗi khung hình
 * — và với một trang bắn request liên tục thì "một vòng nữa mỗi khung hình" là
 * gấp đôi lưu lượng để hiển thị một con số.
 */
export function minutesFromHeaders(
  headers: unknown,
): { remaining: number; limit: number } | null {
  const h = headers as Record<string, string> | undefined;
  if (!h) return null;
  const remaining = Number(h["x-trial-minutes-remaining"]);
  const limit = Number(h["x-trial-minutes-limit"]);
  if (!Number.isFinite(remaining) || !Number.isFinite(limit)) return null;
  return { remaining, limit };
}
